### ISSUE TITLE
RD resolver - fix currently unhandeled individual source errors

### ISSUE THREAD

**Issue by henryjfry** [Tikipeter/tikipeter.github.io#197]
For the source select resolver currently, as it isnt checking the cache status of files, when you select a file which is blocked on real debrid it will not attempt to resolve another file when it fails.

eg  plugin://plugin.video.fenlight/?mode=playback.media&media_type=episode&query=Frasier&year=2001&season=9&episode=6&ep_name=Room Full of Heroes&tmdb_id=3452&premiered=2001-10-30&autoplay=false

This particular show has a bunch of "infringing files" which have been dmca'd on RD and when you select any of them it wont attempt to try the next file.

So it needs a couple of tweaks, the first i think here:

\plugin.video.fenlight\resources\lib\apis\real_debrid_api.py
````
	def resolve_magnet(self, magnet_url, info_hash, store_to_cloud, title, season, episode):
		compare_title = re.sub(r'[^A-Za-z0-9]+', '.', title.replace('\'', '').replace('&', 'and').replace('%', '.percent')).lower()
		elapsed_time, transfer_finished = 0, False
		extensions = supported_video_extensions()
		try:
			torrent = self.add_magnet(magnet_url)
			try: torrent_id = torrent['id']
			except: return None
........
````

So currently I think it has an issue because "torrent_id" cant be populated for an infringing file?
The exception should capture it but is also looking for torrentid?

And then here:
plugin.video.fenlight\resources\lib\modules\sources.py

````
	def play_file(self, results, source={}):
		try: source_index = results.index(source)
		except: source_index = 0
		self.playback_successful, self.cancel_all_playback = None, False
		try:
			hide_busy_dialog()
			url = None
			results = [i for i in results if not 'Uncached' in i.get('cache_provider', '')]
			if not source: source = results[0]
			items = [source]
			if not self.limit_resolve: 
				source_index = results.index(source)
				results.remove(source)
				leading_index = max(source_index-3, 0)
				items_prev = results[leading_index:source_index]
				trailing_index = 7 - len(items_prev)
				items_next = results[source_index:source_index+trailing_index]
				items = items + items_next + items_prev
			processed_items = []
			processed_items_append = processed_items.append
			for count, item in enumerate(items, 1):
				resolve_item = dict(item)
				provider = item['scrape_provider']
				if provider == 'external': provider = item['debrid'].replace('.me', '')
				elif provider == 'folders': provider = item['source']
				provider_text = provider.upper()
				extra_info = '[B]%s[/B] | [B]%s[/B] | %s' %  (item['quality'], item['size_label'], item['extraInfo'])
				display_name = item['display_name'].upper()
				resolve_item['resolve_display'] = '%02d. [B]%s[/B][CR]%s[CR]%s' % (count, provider_text, extra_info, display_name)
				processed_items_append(resolve_item)
				if provider == 'easynews':
					for retry in range(1, 2):
						resolve_item = dict(item)
						resolve_item['resolve_display'] = '%02d. [B]%s (RETRYx%s)[/B][CR]%s[CR]%s' % (count, provider_text, retry, extra_info, display_name)
						processed_items_append(resolve_item)
			items = list(processed_items)
			if not self.continue_resolve_check(): return self._kill_progress_dialog()
			hide_busy_dialog()
			self.playback_percent = self.get_playback_percent()
			if self.playback_percent == None: return self._kill_progress_dialog()
			if not self.resolve_dialog_made: self._make_resolve_dialog()
			if self.background: sleep(1000)
			monitor = xbmc_monitor()
			url = None
			for count, item in enumerate(items, 1):
				try:
					hide_busy_dialog()
					if not self.progress_dialog: break
					self.progress_dialog.reset_is_cancelled()
					self.progress_dialog.update_resolver(text=item['resolve_display'])
					self.progress_dialog.busy_spinner()
					if count > 1:
						sleep(1500)
						try: del player
						except: pass
					player = FenLightPlayer()
					url, self.playback_successful, self.cancel_all_playback = None, None, False
					self.playing_filename = item['name']
					try:
						if self.progress_dialog.iscanceled() or monitor.abortRequested(): break
						url = self.resolve_sources(item)
						if url == None:
							return self.play_file(results, source=results[source_index+1])
						if url:
							resolve_percent = 0
							self.progress_dialog.busy_spinner('false')
							self.progress_dialog.update_resolver(percent=resolve_percent)
							sleep(200)
							player.run(url, self)
						if self.cancel_all_playback: break
						if self.playback_successful: break
						if count == len(items):
							self.cancel_all_playback = True
							player.stop()
							break
					except: pass
				except: pass
		except: self._kill_progress_dialog()
		if self.cancel_all_playback: return self._kill_progress_dialog()
		if not self.playback_successful or not url: self.playback_failed_action()
		try: del monitor
		except: pass
````

So basically when "resolve_sources" returns none it will try the next file and the next file and so on.

**Comment by derrgoo** [comment]
Big help Mr, Henry
